File: mobile/ios/BoxVistaIOS/Services/BoxService.swift
// BoxService.swift
//
//  BoxService.swift
//  BoxVistaIOS
//
//  Created by Pol Linger on 09/06/25.
//

import Foundation
import Combine

/// Protocolo que define la URL base para Boxes
protocol BoxServiceProtocol {
    /// Base URL de la API (ej: http://localhost:3000/api)
    var baseURL: URL { get }
}

extension BoxServiceProtocol {
    /// Obtiene todas las cajas desde la API
    /// - Throws: Error de red o decodificación
    /// - Returns: Array de modelos de dominio `Box`
    func getBoxes() async throws -> [Box] {
        try await withCheckedThrowingContinuation { continuation in
            NetworkManager.shared.fetchBoxes { result in
                switch result {
                case .success(let boxes):
                    continuation.resume(returning: boxes)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Obtiene una caja concreta por ID
    /// - Parameter id: Identificador de la caja
    /// - Throws: Error de red o decodificación
    /// - Returns: Modelo de dominio `Box`
    func getBox(id: Int64) async throws -> Box {
        try await withCheckedThrowingContinuation { continuation in
            NetworkManager.shared.fetchBox(id: Int64(id)) { result in
                switch result {
                case .success(let box):
                    continuation.resume(returning: box)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Crea una nueva caja con sus objetos
    /// - Parameters:
    ///   - name: Nombre de la caja
    ///   - description: Descripción de la caja
    ///   - objects: Array de objetos que contendrá la caja
    /// - Returns: La caja creada
    func createBox(name: String, description: String, objects: [ObjectItem]) async throws -> Box {
        try await withCheckedThrowingContinuation { continuation in
            NetworkManager.shared.createBox(name: name, description: description, objects: objects) { result in
                switch result {
                case .success(let box):
                    continuation.resume(returning: box)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    func deleteBox(box: Box) async throws {
        try await withCheckedThrowingContinuation { continuation in
            NetworkManager.shared.deleteBox(id: box.id) { result in
                switch result {
                case .success:
                    continuation.resume()
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }
        
            
    
}

/// Servicio real que apunta a tu API Express/SQLite
struct BoxService: BoxServiceProtocol {
    let baseURL = URL(string: "http://localhost:3000/api")!
}

/// Servicio de prueba que lee JSON local
struct BoxServiceTest: BoxServiceProtocol {
    let baseURL: URL = {
        Bundle.main.url(forResource: "BoxesTest", withExtension: "json")!
    }()
}

File: mobile/BoxVista/app/src/main/java/com/hirlu/boxvista/services/BoxService.kt
package com.hirlu.boxvista.services

import com.hirlu.boxvista.NetworkManager
import com.hirlu.boxvista.models.Box
import com.hirlu.boxvista.models.ObjectItem

//
interface BoxServiceProtocol {
    suspend fun getBoxes(): List<Box>
    suspend fun getBox(id: Long): Box
    suspend fun createBox(name: String, description: String, objects: List<ObjectItem>): Box
    suspend fun deleteBox(box: Box)
    suspend fun updateBox(box: Box): Box
}

class BoxService : BoxServiceProtocol {

    override suspend fun getBoxes(): List<Box> = NetworkManager.fetchBoxes()

    override suspend fun getBox(id: Long): Box = NetworkManager.fetchBox(id)

    override suspend fun createBox(
        name: String,
        description: String,
        objects: List<ObjectItem>
    ): Box {
        val box = Box(
            name = name,
            description = description,
            objects = objects.toMutableList()
        )
        return NetworkManager.createBox(box)
    }

    override suspend fun deleteBox(box: Box) {
        NetworkManager.deleteBox(box.id!!)
    }

    override suspend fun updateBox(box: Box): Box = NetworkManager.updateBox(box.id!!, box)
}
